edb/pgsql/delta.py

```diff
@@ -4420,35 +4364,177 @@ def _compile_conversion_expr(
                 external_rvars[ptr_path_id, 'source'] = rvar
                 external_rvars[ptr_path_id, 'value'] = rvar
                 external_rvars[src_path_id, 'identity'] = rvar
+                external_rvars[tgt_path_id, 'identity'] = rvar
                 if local_table_only and not is_lprop:
                     external_rvars[src_path_id, 'source'] = rvar
                     external_rvars[src_path_id, 'value'] = rvar
                 elif is_lprop:
-                    external_rvars[tgt_path_id, 'identity'] = rvar
                     external_rvars[tgt_path_id, 'value'] = rvar
             else:
                 src_rvar = compiler.new_external_rvar(
-                    rel_name=(alias,),
+                    rel_name=(source_alias,),
                     path_id=src_path_id,
                     outputs={},
                 )
                 external_rvars[src_path_id, 'identity'] = src_rvar
                 external_rvars[src_path_id, 'value'] = src_rvar
                 external_rvars[src_path_id, 'source'] = src_rvar
-        else:
-            external_rvars = None
 
-        sql_tree = compiler.compile_ir_to_sql_tree(
+        # Wrap the expression into a select with iterator, so DML and
+        # volatile expressions are executed once for each object.
+        #
+        # The result is roughly equivalent to:
+        # for obj in Object union select <expr>
+
+        # generate a unique path id for the outer scope
+        typ = orig_schema.get(f'schema::ObjectType', type=s_types.Type)
+        outer_path = irast.PathId.from_type(
+            orig_schema, typ, typename=sn.QualName("std", "obj")
+        )
+
+        # scope tree wrapping is roughy equivalent to:
+        # "(std::obj) uid:10000": {
+        #   "BRANCH uid:10001",
+        #   "FENCE uid:10002": { ... compiled scope children ... }
+        # }
+        scope_iter = irast.ScopeTreeNode(
+            unique_id=10001,
+        )
+        scope_body = irast.ScopeTreeNode(
+            unique_id=10002,
+            fenced=True
+        )
+        scope_body.children.extend(ir.scope_tree.children)
+
+        scope_root = irast.ScopeTreeNode(
+            unique_id=10000,
+            path_id=outer_path,
+        )
+        scope_root.children.append(scope_iter)
```

Comment:
Don't modify `children` directly, use `attach_child`. Modifying it directly won't set the parent pointer in the children. (Note that `attach_child` will remove it from a current parent, so if you are moving nodes from one tree to another you'll need to copy the list)
(Which I think doesn't matter here, but.)